Toml

rchiossi opened the pull request:
I tried to organize this patch in a way that would make sense to review. The first 4 patches just slowly converts the code to what it will look like once toml lib is used so that the trasition look smoother.

This patch also adds a new `mixer config` subcommand with utilities for validating and converting config files.

kevincwells commented:
First reaction: :open_mouth: +6,613 of code?!. Then remembered vendored dependencies.

rchiossi commented:
Haha indeed. The actual code is about +600 :)

kevincwells commented on mixer/cmd/config.go:
```diff
@@ -89,10 +90,49 @@ and check if all required values were provided`,
+	Long: `Convert an old config file to the new TOML format. The command will generate
+a backup file of the old config and will replace it with the converted one. Environment
+variables will not be expanded and the values will not be validated`,
+	RunE: func(cmd *cobra.Command, args []string) error {
```
**FYI, this doesn't actually work correctly.** You're using the old `builder.go` config parsing to parse, then writing back out with the new `SaveConfig`, but the old `builder.go` config parsing only parsed the parts _it_ cared about. As a result, none of the other sections get converted over.

Example:

```
user@user-clr ~/mix/test $ mixer init 
Creating new builder.conf configuration file...
Initializing mix version 10 from upstream version 21420
Adding bundle "os-core" from upstream bundles
Adding bundle "os-core-update" from upstream bundles
Adding bundle "bootloader" from upstream bundles
Adding bundle "kernel-native" from upstream bundles
```
```
user@user-clr ~/mix/test $ cat builder.conf
[Builder]
SERVER_STATE_DIR=/home/user/mix/test/update
BUNDLE_DIR=/home/user/mix/test/mix-bundles
YUM_CONF=/home/user/mix/test/.yum-mix.conf
CERT=/home/user/mix/test/Swupd_Root.pem
VERSIONS_PATH=/home/user/mix/test

[swupd]
BUNDLE=os-core-update
CONTENTURL=<URL where the content will be hosted>
VERSIONURL=<URL where the version of the mix will be hosted>
FORMAT=1

[Server]
debuginfo_banned=true
debuginfo_lib=/usr/lib/debug/
debuginfo_src=/usr/src/debug/

[Mixer]
LOCAL_BUNDLE_DIR=/home/user/mix/test/local-bundles
```
```
user@user-clr ~/mix/test $ mixer config convert 
user@user-clr ~/mix/test $ cat builder.conf
[Builder]
  BUNDLE_DIR = "/home/user/mix/test/mix-bundles"
  CERT = "/home/user/mix/test/Swupd_Root.pem"
  SERVER_STATE_DIR = "/home/user/mix/test/update"
  VERSIONS_PATH = "/home/user/mix/test"
  YUM_CONF = "/home/user/mix/test/.yum-mix.conf"

[Swupd]
  BUNDLE = ""
  CONTENTURL = ""
  FORMAT = "1"
  VERSIONURL = ""

[Server]
  DEBUG_INFO_BANNED = ""
  DEBUG_INFO_LIB = ""
  DEBUG_INFO_SRC = ""

[Mixer]
  LOCAL_BUNDLE_DIR = "/home/user/mix/test/local-bundles"
  LOCAL_REPO_DIR = ""
  LOCAL_RPM_DIR = ""
```

rchiossi replied:
Yep, I'm I noticed it today. I'm updating the old parser to include the missing variables. Regarding where the code lies, I dodn't expect the convert code do live long, so I didn't see the need to add an exported method in the config object. But I see your point.

kevincwells replied:
I agree that this code probably won't survive all that long, and given that, it is indeed probably easiest to just extend the existing regex matching like you're doing.  If we needed INI parsing to persist for longer, it'd be better to use the `go-ini/ini` package the way `builder/bundles.go` is currently doing.

But hopefully no one says we need to support the old config for too long.

kevincwells replied:
Haha, that's true. But just for the record, I'm not suggesting you do that unless we have some pressing need to keep the old config support around for a long time. :)

kevincwells commented on builder/config.go:
```diff
@@ -23,38 +23,133 @@ import (
+		return config.createLegacyConfig(localrpms)
+	}
+
+	if err := config.LoadDefaults(); err != nil {
```
You've lost `localrpms` here, which means running `mixer init --local-rpms --new-config` still leaves those two variables as empty strings.

rchiossi replied:
nice catch.

kevincwells commented on builder/config.go:
```diff
@@ -158,3 +155,47 @@ func (config *MixConfig) LoadConfig(filename string) error {
+
+func (config *MixConfig) expandEnv() error {
+	re := regexp.MustCompile(`\$\{?([[:word:]]+)\}?`)
+	rc := reflect.ValueOf(config).Elem()
```
Not entirely sure why this comment is being marked as outdated on GitHub. Tagging @rchiossi just to make sure he sees it.

rchiossi replied:
my understanding is that they yield the same result. The difference is just the day I wrote the code :)
I can change for consistency.

rchiossi commented on builder/config.go:
```diff
@@ -158,3 +155,47 @@ func (config *MixConfig) LoadConfig(filename string) error {
+
+	for i := 0; i < rc.NumField(); i++ {
+		s := rc.Field(i).Addr().Interface()
+		section := reflect.Indirect(reflect.ValueOf(s))
```
Nop, it doesn't work. Without all this code you get an object that is readable, but not writable.

kevincwells replied:
Ah, right, because you created `rc` with `reflect.ValueOf(config).Elem()`, so it's not "settable". If you instead were to create it with the original reference (`reflect.ValueOf(config)`), you'd have a access to settable `Value`'s.

rchiossi replied:
Nop, also doesn't work. The problem here is not to get settable references to the members of the base structure, but to get settable references to the substructures within.

rchiossi replied:
Actually, it turns out that reflect.ValueOf(config).Elem() allows for settable substructures while reflect.ValueOf(*config) does not. After some testing with different combinations I managed to scrap some of the convoluted reflection code.

rchiossi commented on builder/config.go:
```diff
@@ -158,3 +155,47 @@ func (config *MixConfig) LoadConfig(filename string) error {
+
+	for i := 0; i < rc.NumField(); i++ {
+		sectionT := reflect.TypeOf(rc.Field(i).Interface())
+		sectionV := reflect.ValueOf(rc.Field(i).Interface())
```
For this case it might actually work. I'll give it a try.

rchiossi commented on mixer/cmd/config.go:
```diff
@@ -0,0 +1,105 @@
+	Short: "Perform config related actions",
+}
+
+func initConfig() error {
```
Yep I agree. I was thinking about puting it on helpers instead, or keep the function agnostic of the name since config doesn't know b.BuildConfFile

kevincwells replied:
I don't think it should go in `helpers.go`, because it's a builder concept; it should be in the `builder` package.

rchiossi replied:
it is only a builders concept if you include the "builder.conf" string. I intend to have it as a parameter.

rchiossi commented on mixer/cmd/config.go:
```diff
@@ -0,0 +1,105 @@
+	Short: "Parse a config file and print its properties",
+	Long: `Parse a builder config file and display its properties. Properties containing
+environment variables will be expanded`,
+	RunE: func(cmd *cobra.Command, args []string) error {
```
Indeed. Just a copy and paste left over on the RunE.

rchiossi commented on mixer/cmd/config.go:
```diff
@@ -0,0 +1,105 @@
+	Short: "Parse a config file and return an error if it is not valid",
+	Long: `Parse a config file, expand any environment variable defined in the file
+and check if all required values were provided`,
+	RunE: func(cmd *cobra.Command, args []string) error {
```
ack

rchiossi commented:
> I would remove the mixer config validate command and just rename mixer config read to mixer config validate. Right now they're equivalent (with one just being the silent version of the other), and the intention of read (i.e., "parse and fully expand the config, then output it to the user to read and verify") really is a validation step.
> I also hate validate commands that only spit out the first error they hit, but I understand that's a limitation of the toml parsing library.

At first I though it was a good idea to add `validate` as a separate command to make it easier to use it in scripts, but I guess it is just as easy to redirect the output of `read`. I'll unify them and keep the `validate` name. As for the errrs, indeed that is a limitation of the lib.

> The output of mixer config read seems a bit odd to me -- it looks a lot like a TOML file, but it isn't a valid TOML file. It's supposed to just be a visual representation of our internal, fully-parsed data structure, right? Maybe just drop the [] around the section titles? Or go hard the other direction and output valid TOML.
I agree. I believe it is better to just output TOML then.

> Do we actually want to add a mixer config convert command? I think providing a conversion tool (once you get it working) is a great idea, but we've been talking a lot lately about being careful of what commands we introduce for API compatibility reasons. A year or two from now, when people forget there ever was a non-TOML config, do we want this command hanging around? Are some commands safe to remove? Do we perhaps always want to have an evolving config convert command that knows how to translate from the previous config format to the current? (We've talked about needing to convert CLI commands, but not the config file itself)

As I understand, the problem with adding another command and then removing it is to potentially break some automation tool that uses it. For that, `mixer convert` or a separate tool `mixer-convert` would cause the same harm once they are removed. It just seemed simpler to add it as an extra command. Either way, if we provide that facility we will need to remove it once we drop the old config. Having an evolving convert command can be really tricky since we don't have versioning for our config file. Maybe it is a good idea to add a `CONFIG_VERSION=xx` to the `Mixer` section going forward.


> Update bundles.go so that it actually uses this new parsing.
you mean, drop the old config?

> Move mixversion, mixbundles, upstreamversion, and upstreamurl into the new TOML builder.conf?
Yeah, I think those values make more sense in the config than in a separate file.

> Add individual config getters and setters? (I'm thinking as analogous to get config <field> <value> and git config --get <field>.) TOML is pretty easy to edit, so this might not add a lot of value on top of the full-file getter mixer bundle read. (This becomes more relevant as config files get larger, so you don't have to parse through a long document to see the value of a single config field.)
you mean `mixer config get` and `mixer config set`? Actually that is what I was writing today! :)

> Refactor/overhaul the config file content!!! :D
That is a bit trickier, but adding a version to the config might make our lives easier here.

kevincwells commented:
>> Update bundles.go so that it actually uses this new parsing.

> you mean, drop the old config?

If by "old config" you mean the ini parsing it's doing. If you mean the old stuff that was in `builder.go`, then no.  Even now, `builder/bundles.go` still uses `github.com/go-ini/ini` to pull all of the values it cares about (except for the overall build format) into a `cfg` object in `readBuildBundlesConfig`. Now that you're reading in all of the values in the main config, this entire chunk of logic can be removed. (In fact, it has `// TODO: Move this to the more general configuration handling.` at the top :wink: )

kevincwells commented:
>> Refactor/overhaul the config file content!!! :D

> That is a bit trickier, but adding a version to the config might make our lives easier here.

and

> Having an evolving convert command can be really tricky since we don't have versioning for our config file. Maybe it is a good idea to add a CONFIG_VERSION=xx to the Mixer section going forward.

That's a good idea. The implementation of it is tricky if we really want it to be format-agnostic (i.e., we'd want to be able to check the format version regardless of the format encoding itself, be it INI or TOML or JSON).

rchiossi commented on builder/config.go:
```diff
@@ -199,3 +199,20 @@ func (config *MixConfig) validate() error {
+// Print print variables and values of a MixConfig struct
+func (config *MixConfig) Print() {
+	rt := reflect.TypeOf(*config)
+	rv := reflect.ValueOf(*config)
```
it was originally called rc here as well, but since this is the only place I needed rt, I changed it. I think it is better to just change the other to rv instead.

rchiossi commented on builder/config.go:
```diff
@@ -199,3 +199,20 @@ func (config *MixConfig) validate() error {
+		fmt.Printf("[%s]\n", rt.Field(i).Name)
+
+		for j := 0; j < sectionT.NumField(); j++ {
+			fmt.Printf("  .%s = %s\n", sectionT.Field(j).Name, sectionV.Field(j).String())
```
This was intentional, it is just an indication of a member of this object. But this will be gone with a toml output.

rchiossi commented:
I updated the PR with the changes requested by @kevincwells 
- Moved convert functionality from cmd to config
- Created a GetConfigPath function in builder
- Renamed `read` to `validate` and removed the old `validate`
- Changed `mixer config read` to output TOML
- Fixed the minor code issues and cleaned up a bit.

rchiossi commented:
Hmm, apparently there is a rebase conflict. I'll update in a minute.

tmarcu commented on builder/config.go:
```diff
@@ -0,0 +1,358 @@
+
+	builderconf := filepath.Join(pwd, "builder.conf")
+
+	err = helpers.CopyFileNoOverwrite(builderconf, "/usr/share/defaults/bundle-chroot-builder/builder.conf")
```
For future patch: We should really update this file itself to the new format so the user doesn't have to call convert on it later, and probably move it out of BCB and into mixer @matthewrsj

matthewrsj replied:
I'd call the file under bcb deprecated. Does Clear devops still use this or something?

rchiossi replied:
I thought we were dropping that file since `mixer init --new-config` will generate the template in the new format already.

kevincwells replied:
Agreed to both. If I understand correctly, it's only used in the legacy code path, so rather than update the file we can just eventually remove it when `--new-config` is removed.

matthewrsj commented on builder/bundles.go:
```diff
@@ -609,7 +609,7 @@ lib=%s
 `, cfg.DebugInfoBanned, cfg.DebugInfoLib, cfg.DebugInfoSrc)
 	}
-	err = ioutil.WriteFile(filepath.Join(b.StateDir, "server.ini"), serverINI.Bytes(), 0644)
+	err = ioutil.WriteFile(filepath.Join(b.Config.Builder.ServerStateDir, "server.ini"), serverINI.Bytes(), 0644)
```
Future patch, `build update` should just call the original conf instead of relying on this. This isn't a change required for this PR though.

rchiossi replied:
Yep, the idea is to get rid of the other conf files. I hope to get this on with my next patch.

tmarcu commented on builder/config.go:
```diff
@@ -0,0 +1,358 @@
+}
+
+// Convert parses an old config file and converts it to TOML format
+func (config *MixConfig) Convert(filename string) error {
```
My only nitpick so far...calling convert again borks the builder.conf if it has already been converted/is in any format other than the original format. Granted, I don't expect people to run it twice, but you never know. It might be nice to see if we can already read it in the new format and just return nil in that case.

rchiossi replied:
That is one of the reasons I was talking about adding a version number, so we can identify what config we are parsing. Is this case it won't really be a problem since calling convert on a new config will fail on the first parse step.

kevincwells replied:
Agreed to both. Good simple next step :)
